### PR TITLE
[IMP][website_forum] forum improvements

### PR DESCRIPTION
**Pull request by rga-odoo**
Task: https://www.odoo.com/web#id=10387&view_type=form&model=project.task&action=333&active_id=131
Pad: https://pad.odoo.com/p/openerp-project.task-MAKW3FS9Y3

[IMP][website_forum] improved logic for posting links in forum + generated karma list dynamically
[IMP][website_forum] enabled edit feature for links, added char pager for tags
[IMP][website_forum] solved blog's css conflict
[FIX][website_forum] added login redirection in ask questions
[IMP][wenbsite_forum] improved help for relevancey formula
[FIX][website_forum] solved counter problem for vote & solved mysterious view counter problem (generated because of survey)

### CODE REVIEW

File: addons/survey/static/src/js/survey.js [hunk 1]
```diff
@@ -145,7 +145,9 @@ openerp.website.if_dom_contains('.js_surveyform', function (the_form) {
     // });
 
     // Launch prefilling
-    prefill();
+    if(prefill_controller){
+        prefill();
+    }
```
Comment (jem-odoo) [inline addons/survey/static/src/js/survey.js]: Why this changes ?
I think the commit fix your problem (if there was one) : https://github.com/odoo/odoo/commit/11f147f2d94c8aff729590183e24337b2c076c86

File: addons/website_forum/views/website_forum.xml [hunk 2]
```diff
@@ -192,6 +199,12 @@
     </t>
 </template>
 
+<template id="introduction_message">
```
Comment (jem-odoo) [inline addons/website_forum/views/website_forum.xml]: Add a comment explaining this is to edit the welcome message

File: addons/website_forum/views/website_forum.xml [hunk 3]
```diff
@@ -258,6 +272,45 @@
 <template id="faq">
     <t t-call="website_forum.header">
         <div t-field="forum.faq"/>
+
+<table class="table table-striped">
```
Comment (jem-odoo) [inline addons/website_forum/views/website_forum.xml]: Ident the code and add a title explaining the contain of the table

File: addons/website_forum/controllers/main.py [hunk 4]
```diff
@@ -4,7 +4,8 @@
 import werkzeug.wrappers
 import simplejson
 import lxml
-from urllib2 import urlopen
+import urllib
+from urllib2 import urlopen,URLError
```
Comment (jem-odoo) [inline addons/website_forum/controllers/main.py]: please order your import : 
import A
import B
import C
from v import W
from X import Y

;)

File: addons/website_forum/models/forum.py [hunk 5]
```diff
@@ -44,11 +44,22 @@ def _get_default_faq(self):
...
     faq = fields.Html('Guidelines', default=_get_default_faq, translate=True)
-    description = fields.Html(
+    description = fields.Text(
         'Description',
-        default='<p> This community is for professionals and enthusiasts of our products and services.'
+        default='This community is for professionals and enthusiasts of our products and services.'
                 'Share and discuss the best content and new marketing ideas,'
-                'build your professional profile and become a better marketer together.</p>')
+                'build your professional profile and become a better marketer together.')
+    introduction_message = fields.Html(
```
Comment (jem-odoo) [inline addons/website_forum/models/forum.py]: keep the new field, but call it `welcome_message`.

File: addons/website_forum/models/forum.py [hunk 6]
```diff
@@ -45,11 +45,22 @@ def _get_default_faq(self):
...
+        'Introduction Message',
+        default = """<section class="bg-info" style="height: 168px;"><div class="container">
```
Comment (jem-odoo) [inline addons/website_forum/models/forum.py]: After a little chat with @tde-banana-odoo , we decided to keep the default value harcoded like this (even) it is not so beautiful. Create a new forum is a rare activity, so it can stay like that, instead of a customizable template.
Maybe on day, we will change this, but not for now. Thanks.

File: addons/website_forum/static/src/js/website_forum.editor.js [hunk 7]
```diff
@@ -9,10 +9,24 @@
...
+                        '</div>');
+                    $add.find('label').append(_t("Add page in menu"));
+                    $group.after($add);
+                }
+            }).then(function (forum_name, field, $dialog) {
+                var add_menu = ($dialog.find('input[type="checkbox"]').is(':checked'));
                 website.form('/forum/new', 'POST', {
-                    forum_name: forum_name
+                    forum_name: forum_name,
+                    add_menu: add_menu || ""
```
Comment (jem-odoo) [inline addons/website_forum/static/src/js/website_forum.editor.js]: set false instead of ""

File: addons/website_forum/static/src/js/website_forum.js [hunk 8]
```diff
@@ -6,8 +8,9 @@
                 ev.preventDefault();
                 var $warning = $('<div class="alert alert-danger alert-dismissable oe_forum_alert" id="karma_alert">'+
                     '<button type="button" class="close notification_close" data-dismiss="alert" aria-hidden="true">&times;</button>'+
-                    karma + ' karma is required to perform this action. You can earn karma by having '+
-                            'your answers upvoted by the community.</div>');
+                    karma + _t(' karma is required to perform this action. You can earn karma by having '+
+                            'your answers upvoted by the community.')+
+                            '</div>');
```
Comment (jem-odoo) [inline addons/website_forum/static/src/js/website_forum.js]: make it only one string instead of 'blablala' + 'blablala'

File: addons/website_forum/models/forum.py [hunk 9]
```diff
@@ -128,8 +139,27 @@ def _tag_to_write_vals(self, tags=''):
         post_tags.insert(0, [6, 0, existing_keep])
         return post_tags
 
+    @api.one
+    @api.constrains('allow_link','allow_question','allow_discussion','default_post_type')
```
Comment (jem-odoo) [inline addons/website_forum/models/forum.py]: use a linter : 1 space after the comma

File: addons/website_forum/static/src/js/website_forum.js [hunk 10]
```diff
@@ -119,30 +123,46 @@
...
+                if(data){
+                    $form.find("input[name='post_name']").val(data);
+                    $form.find("button#btn_post_your_article").prop('disabled',false);
+                    $link.val(url);
+                    $('.invalid_url_warning').remove();
+                    $form.find(".title_loading").removeClass('fa-spin fa-cog').addClass('fa-check');
+                }else{
+                    var $warning = $('<div class="alert alert-danger alert-dismissable invalid_url_warning" style="position:absolute; margin-top: -180px; margin-left: 90px;">'+
+                        '<button type="button" class="close notification_close" data-dismiss="alert" aria-hidden="true">&times;</button>'+
+                        _t('Please enter valid URl.')+
```
Comment (jem-odoo) [inline addons/website_forum/static/src/js/website_forum.js]: URL, with a capital L

File: addons/website_forum/controllers/main.py [hunk 11]
```diff
@@ -164,23 +172,40 @@ def tag_read(self, q='', l=25, **post):
         )
         return simplejson.dumps(data)
 
-    @http.route(['/forum/<model("forum.forum"):forum>/tag'], type='http', auth="public", website=True)
-    def tags(self, forum, page=1, **post):
-        tags = request.env['forum.tag'].search([('forum_id', '=', forum.id), ('posts_count', '>', 0)], limit=None, order='posts_count DESC')
+    @http.route(['/forum/<model("forum.forum"):forum>/tag','/forum/<model("forum.forum"):forum>/tag/<tag_char>'], type='http', auth="public", website=True)
+    def tags(self, forum, page=1, tag_char=False, **post):
```
Comment (jem-odoo) [inline addons/website_forum/controllers/main.py]: tag_char shouldn't be 'All' by default ?

File: addons/website_forum/controllers/main.py [hunk 12]
```diff
@@ -164,23 +172,40 @@ def tag_read(self, q='', l=25, **post):
...
+        pager_chars = forum.get_tags_char()
+        tag_char = urllib.unquote(tag_char).decode("unicode_escape") if tag_char else (pager_chars[0][0] if pager_chars else 'All')
+        if pager_chars: pager_chars.insert(0,('All',0))
```
Comment (jem-odoo) [inline addons/website_forum/controllers/main.py]: add a line break, and move this condition before your decoding stuff

File: addons/website_forum/controllers/main.py [hunk 13]
```diff
@@ -164,23 +172,40 @@ def tag_read(self, q='', l=25, **post):
         )
         return simplejson.dumps(data)
 
-    @http.route(['/forum/<model("forum.forum"):forum>/tag'], type='http', auth="public", website=True)
-    def tags(self, forum, page=1, **post):
-        tags = request.env['forum.tag'].search([('forum_id', '=', forum.id), ('posts_count', '>', 0)], limit=None, order='posts_count DESC')
+    @http.route(['/forum/<model("forum.forum"):forum>/tag','/forum/<model("forum.forum"):forum>/tag/<tag_char>'], type='http', auth="public", website=True)
+    def tags(self, forum, page=1, tag_char=False, **post):
+        pager_chars = forum.get_tags_char()
+        tag_char = urllib.unquote(tag_char).decode("unicode_escape") if tag_char else (pager_chars[0][0] if pager_chars else 'All')
```
Comment (jem-odoo) [inline addons/website_forum/controllers/main.py]: I need to be enlighten : why are you decoding this ?

File: addons/website_forum/controllers/main.py [hunk 14]
```diff
@@ -164,23 +172,40 @@ def tag_read(self, q='', l=25, **post):
...
+        domain = [('forum_id', '=', forum.id), ('posts_count', '>', 0)]
+        if tag_char.lower() != "all":
+            domain.append(('name','=ilike',tag_char+'%'))
+        tags = request.env['forum.tag'].search(domain, limit=None, order='posts_count DESC' if tag_char=='All' else 'name')
         values = self._prepare_forum_values(forum=forum, searches={'tags': True}, **post)
         values.update({
             'tags': tags,
             'main_object': forum,
+            'pager_chars': pager_chars,
+            'active_tag':tag_char,
```
Comment (jem-odoo) [inline addons/website_forum/controllers/main.py]: transform it into an upper letter, otherwise, if you compare the active_char with the keys of pager_chars, it can not match, if a lowercase is introduce in the url.

File: addons/website_forum/controllers/main.py [hunk 15]
```diff
@@ -164,23 +172,40 @@ def tag_read(self, q='', l=25, **post):
         )
         return simplejson.dumps(data)
 
-    @http.route(['/forum/<model("forum.forum"):forum>/tag'], type='http', auth="public", website=True)
-    def tags(self, forum, page=1, **post):
-        tags = request.env['forum.tag'].search([('forum_id', '=', forum.id), ('posts_count', '>', 0)], limit=None, order='posts_count DESC')
+    @http.route(['/forum/<model("forum.forum"):forum>/tag','/forum/<model("forum.forum"):forum>/tag/<tag_char>'], type='http', auth="public", website=True)
```
Comment (jem-odoo) [inline addons/website_forum/controllers/main.py]: the tag_char is not typed : try with `<string(length=1)`. Give it a try ;)

File: addons/website_forum/controllers/main.py (hunk 13 above)
Comment (pga-odoo) [inline addons/website_forum/controllers/main.py]: It is because of non-ascii characters.Like Chinese, Hindi, Latin etc
And that's why i need to use decoding and why i can not use <string(length=1)
download

File: addons/website_forum/controllers/main.py (hunk 11 above)
Comment (pga-odoo) [inline addons/website_forum/controllers/main.py]: Currently in master there is no such filter so it loads all tags and if there are lot's of tags(like we have more than 4800 tags in odoo help) then it consume to much time.
So currentlly it is taking first possible char by default.

File: addons/website_forum/models/forum.py (hunk 6 above)
Comment (pga-odoo) [inline addons/website_forum/models/forum.py]: I'm little-bit confused here because in previous comment you just said to rename this field.
So what to do. Revert it like master or only remove edition feature

File: addons/website_forum/static/src/js/website_forum.editor.js (hunk 7 above)
Comment (pga-odoo) [inline addons/website_forum/static/src/js/website_forum.editor.js]: It is blank because in controller it is sending false as string

File: addons/website_forum/views/website_forum.xml [hunk 16]
```diff
@@ -369,8 +429,9 @@
             We keep a high level of quality in showcased posts, around 20% of the submited
             posts will be featured.
         </p>
-        <form t-attf-action="/forum/#{ slug(forum) }/new?post_type=link" method="post" role="form" class="tag_text form-horizontal">
+        <form t-att-action="action_url" method="post" role="form" class="post_link tag_text form-horizontal">
```
Comment (tde-banana-odoo) [inline addons/website_forum/views/website_forum.xml]: Why adding a post_link class that does not seem to be used anywhere ?

File: addons/website_forum/views/website_forum.xml [hunk 17]
```diff
@@ -360,7 +412,15 @@
 <!-- Edition: Post Article -->
 <template id="new_link">
     <t t-call="website_forum.header">
-        <h1 class="mt0">Submit a Link</h1>
+        <t t-set="link_create" t-value="False if post_name else True"/>
```
Comment (tde-banana-odoo) [inline addons/website_forum/views/website_forum.xml]: ? What is the purpose of that ?

Comment (pga-odoo) [inline addons/website_forum/views/website_forum.xml]: Before this it is not possible to edit previously posted Link.
But now with this you can make change in previously posted links. 
I'm using same create form for edit links. Just changing action_url and string to 'Edit your Link' from 'Submit a Link' This `link_create` is holding a state weather this is edit or create

Comment (jem-odoo) [inline addons/website_forum/views/website_forum.xml]: fixed at eb624313177ac7183e16b8c8be5044213fd81fb8

File: addons/website_forum/views/website_forum.xml (hunk 16 above)
Comment (jem-odoo) [inline addons/website_forum/views/website_forum.xml]: fixed at eb624313177ac7183e16b8c8be5044213fd81fb8
